code/FOV QC/FOV QC utils.R

```diff
@@ -0,0 +1,297 @@
+    names_of_bits_to_plot  <- paste0(rep(flaggedreportercycles, each = 4), rep(c("B", "G", "R","Y"), length(flaggedreportercycles)))
+    bits_to_plot <- match(names_of_bits_to_plot, colnames(res$resid))
+  }
+  if (bits == "flagged_bits") {
+    bits_to_plot  <- which(colSums(res$fovstats$flag) > 0)
+  }
+  if (bits == "all") {
+    bits_to_plot <- 1:ncol(res$resid)
+  }
+  temp <- sapply(bits_to_plot, function(i) {
+    if (!is.null(outdir)) {
+      png(paste0(outdir, "/", make.names(colnames(res$resid)[i]), ".png"), width = plotwidth, height = plotheight, units = "in", res = 300)
+    }
+    par(mar = c(0,0,2,0))
+    plot(res$xy, cex = 0.2, asp = 1, pch = 16,
+         col = colorRampPalette(c("darkblue", "blue", "grey80", "red", "darkred"))(101)[
+           pmax(pmin(51 + res$resid[match(res$gridinfo$gridid, rownames(res$resid)), i] * 50, 101), 1)], 
+         main = paste0(colnames(res$resid)[i], ": log2(fold-change)\nfrom comparable regions elsewhere"))
+    for (f in unique(res$fov)) {
+      inds <- res$fov == f
+      rect(min(xy[inds, 1]), min(xy[inds, 2]), max(xy[inds, 1]), max(xy[inds, 2]))
+    }
+    legend("right", pch = 16,
+           col = rev(c("darkblue", "blue", "grey80", "red", "darkred")),
+           legend = rev(c("< -1", -0.5, 0, 0.5, "> 1")))
+    if (!is.null(outdir)) {
+      dev.off()
+    }
+  })
+}
+
+
+#' Heatmap of estimated bit bias across FOVs
+#' 
+#' @param res Results object created by runFOVQC
+#' @return Draws a heatmap
+#' @export
+FOVEffectsHeatmap <- function(res) {
+  pheatmap::pheatmap(res$fovstats$bias * (res$fovstats$flag),
+                     col = colorRampPalette(c("darkblue", "blue", "white", "red", "darkred"))(100),
+                     breaks = seq(-2,2,length.out = 101),
+                     main = "FOV bias: log2(fold-change) from comparable regions in other FOVs")
+}
+
+
+
+
+#' Get nearest neighbors, sampling diffusely across other FOVs. 
+#' 
+#' @param x Data matrix, obs in rows, variables in columns
+#' @param fov vector of FOV IDs, aligned to rows of x
+#' @param n_neighbors How many neighbors to record
+#' @return A matrix of dimension nrow(obs) * n_neighbors, giving the row IDs for each obs's selected neighbors.
+getNearestNeighborsByFOV <- function(x, fov, n_neighbors = 10) {
+  # get nearest:
+  topneighbors <- FNN::get.knnx(data = x, 
+                                query = x, 
+                                k = min(n_neighbors * 50, nrow(x)))$nn.index
+  # reduce neighbors: none from own FOV, max of max_per_fov from each other fov:
+  reducedneighbors <- t(sapply(1:nrow(topneighbors), function(i) {
+    tempneighbors <- topneighbors[i, ]
+    # only accept 1 per FOV
+    tempneighbors[duplicated(fov[topneighbors[i, ]])] <- NA
+    # none from own FOV:
+    tempneighbors[fov[topneighbors[i, ]] == fov[i]] <- NA
+    return(tempneighbors[!is.na(tempneighbors)][1:n_neighbors])
+  }))
+  
+  return(reducedneighbors)
+}
+
+
+#' Get gridded expression within FOVs
```

Comment:
This makes grids but doesn't get expression per grid. Change description.

Reply:
done